File: backend/app/schemas/period.py
"""
Period Control Schemas

Pydantic schemas for period control and finalization API.
"""
from datetime import datetime, date
from typing import Optional, List, Dict, Any
from uuid import UUID
from decimal import Decimal
from pydantic import BaseModel, Field
from enum import Enum


class PeriodStatus(str, Enum):
    """Status of an accounting period."""
    OPEN = "OPEN"
    REVIEW = "REVIEW"
    FINALIZED = "FINALIZED"
    LOCKED = "LOCKED"


class PeriodResponse(BaseModel):
    """Response for an accounting period."""
    id: UUID
    administration_id: UUID
    name: str
    period_type: str
    start_date: date
    end_date: date
    status: PeriodStatus
    is_closed: bool
    # Timestamps
    created_at: datetime
    closed_at: Optional[datetime] = None
    review_started_at: Optional[datetime] = None
    finalized_at: Optional[datetime] = None
    locked_at: Optional[datetime] = None
    # Who performed actions
    review_started_by_id: Optional[UUID] = None
    finalized_by_id: Optional[UUID] = None
    locked_by_id: Optional[UUID] = None

    class Config:
        from_attributes = True


class ValidationIssue(BaseModel):
    """A validation issue preventing or warning about finalization."""
    id: str
    code: str
    title: str
    description: str
    suggested_action: Optional[str] = None


class ValidationStatus(BaseModel):
    """Validation status for a period."""
    red_issues: List[ValidationIssue]
    yellow_issues: List[ValidationIssue]
    can_finalize: bool
    validation_summary: Dict[str, int]


class PeriodWithValidationResponse(BaseModel):
    """Period with its validation status."""
    period: PeriodResponse
    validation: ValidationStatus


class ReviewPeriodRequest(BaseModel):
    """Request to start period review."""
    notes: Optional[str] = Field(None, description="Optional notes about the review")


class ReviewPeriodResponse(BaseModel):
    """Response from starting period review."""
    period: PeriodResponse
    validation_run_id: UUID
    issues_found: int
    message: str


class FinalizePeriodRequest(BaseModel):
    """Request to finalize a period."""
    acknowledged_yellow_issues: Optional[List[str]] = Field(
        None, 
        description="List of YELLOW issue IDs that have been acknowledged"
    )
    notes: Optional[str] = Field(None, description="Optional notes about finalization")


class FinalizePeriodResponse(BaseModel):
    """Response from finalizing a period."""
    period: PeriodResponse
    snapshot_id: UUID
    message: str


class LockPeriodRequest(BaseModel):
    """Request to lock a finalized period."""
    notes: Optional[str] = Field(
        None, 
        description="Optional notes about why the period is being locked"
    )
    confirm_irreversible: bool = Field(
        False, 
        description="Must be true to confirm understanding that locking is irreversible"
    )


class LockPeriodResponse(BaseModel):
    """Response from locking a period."""
    period: PeriodResponse
    message: str


class SnapshotSummary(BaseModel):
    """Summary metrics from a period snapshot."""
    total_assets: Optional[Decimal] = None
    total_liabilities: Optional[Decimal] = None
    total_equity: Optional[Decimal] = None
    net_income: Optional[Decimal] = None
    total_ar: Optional[Decimal] = None
    total_ap: Optional[Decimal] = None
    vat_payable: Optional[Decimal] = None
    vat_receivable: Optional[Decimal] = None


class PeriodSnapshotResponse(BaseModel):
    """Response for a period snapshot."""
    id: UUID
    period_id: UUID
    administration_id: UUID
    snapshot_type: str
    created_at: datetime
    created_by_id: Optional[UUID] = None
    # Summary metrics
    summary: SnapshotSummary
    # Full report data (JSONB)
    balance_sheet: Optional[Dict[str, Any]] = None
    profit_and_loss: Optional[Dict[str, Any]] = None
    vat_summary: Optional[Dict[str, Any]] = None
    open_ar_balances: Optional[Dict[str, Any]] = None
    open_ap_balances: Optional[Dict[str, Any]] = None
    trial_balance: Optional[List[Dict[str, Any]]] = None
    # Issue acknowledgments
    acknowledged_yellow_issues: Optional[List[str]] = None
    issue_summary: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class AuditLogResponse(BaseModel):
    """Response for a period audit log entry."""
    id: UUID
    period_id: UUID
    administration_id: UUID
    action: str
    from_status: Optional[str] = None
    to_status: str
    performed_by_id: UUID
    performed_at: datetime
    notes: Optional[str] = None
    snapshot_id: Optional[UUID] = None

    class Config:
        from_attributes = True


class PeriodAuditLogsResponse(BaseModel):
    """Response for period audit logs list."""
    period_id: UUID
    logs: List[AuditLogResponse]
    total_count: int


class PeriodsListResponse(BaseModel):
    """Response for list of periods."""
    administration_id: UUID
    periods: List[PeriodResponse]
    total_count: int


class FinalizationPrerequisiteErrorResponse(BaseModel):
    """Error response when finalization prerequisites are not met."""
    detail: str
    red_issues: List[ValidationIssue]
    yellow_issues: List[ValidationIssue]

File: src/lib/api/period.ts
// ============ Period Control API (types + functions) ============
// Extracted from src/lib/api.ts as part of the api.ts decomposition.
// Behavior, endpoints, and response shapes are unchanged.

import { api } from '../api'

// ============ Period Control Types ============

export type PeriodStatus = 'OPEN' | 'REVIEW' | 'FINALIZED' | 'LOCKED' | 'READY_FOR_FILING'

export interface Period {
  id: string
  administration_id: string
  name: string
  period_type: string
  start_date: string
  end_date: string
  status: PeriodStatus
  is_closed: boolean
  created_at: string
  closed_at: string | null
  review_started_at: string | null
  finalized_at: string | null
  locked_at: string | null
  review_started_by_id: string | null
  finalized_by_id: string | null
  locked_by_id: string | null
}

export interface ValidationIssue {
  id: string
  code: string
  title: string
  severity: 'RED' | 'YELLOW'
}

export interface ValidationStatus {
  red_issues: ValidationIssue[]
  yellow_issues: ValidationIssue[]
  can_finalize: boolean
  validation_summary: string
}

export interface PeriodWithValidation {
  period: Period
  validation: ValidationStatus
}

export interface PeriodsListResponse {
  administration_id: string
  periods: Period[]
  total_count: number
}

export interface ReviewPeriodRequest {
  notes?: string
}

export interface ReviewPeriodResponse {
  period: Period
  validation_run_id: string
  issues_found: number
  message: string
}

export interface FinalizePeriodRequest {
  acknowledged_yellow_issues?: string[]
  notes?: string
}

export interface FinalizePeriodResponse {
  period: Period
  snapshot_id: string
  message: string
}

export interface LockPeriodRequest {
  confirm_irreversible: boolean
  notes?: string
}

export interface LockPeriodResponse {
  period: Period
  message: string
}

export interface UpdatePeriodStatusRequest {
  status: 'READY_FOR_FILING' | 'FINALIZED'
}

export interface UpdatePeriodStatusResponse {
  period: Period
  message: string
}

export interface SnapshotSummary {
  total_assets: number
  total_liabilities: number
  total_equity: number
  net_income: number
  total_ar: number
  total_ap: number
  vat_payable: number
  vat_receivable: number
}

export interface PeriodSnapshot {
  id: string
  period_id: string
  administration_id: string
  snapshot_type: string
  created_at: string
  created_by_id: string
  summary: SnapshotSummary
  balance_sheet: Record<string, unknown> | null
  profit_and_loss: Record<string, unknown> | null
  vat_summary: Record<string, unknown> | null
  open_ar_balances: Record<string, unknown> | null
  open_ap_balances: Record<string, unknown> | null
  trial_balance: Record<string, unknown> | null
  acknowledged_yellow_issues: string[] | null
  issue_summary: Record<string, unknown> | null
}

export interface AuditLogEntry {
  id: string
  period_id: string
  administration_id: string
  action: string
  from_status: string | null
  to_status: string | null
  performed_by_id: string
  performed_at: string
  notes: string | null
  snapshot_id: string | null
}

export interface PeriodAuditLogsResponse {
  period_id: string
  logs: AuditLogEntry[]
  total_count: number
}

// Period Control API
export const periodApi = {
  listPeriods: async (
    clientId: string, 
    status?: PeriodStatus[]
  ): Promise<PeriodsListResponse> => {
    const params: Record<string, unknown> = {}
    if (status && status.length > 0) {
      params.status = status
    }
    const response = await api.get<PeriodsListResponse>(
      `/accountant/clients/${clientId}/periods`,
      { params }
    )
    return response.data
  },

  getPeriod: async (clientId: string, periodId: string): Promise<PeriodWithValidation> => {
    const response = await api.get<PeriodWithValidation>(
      `/accountant/clients/${clientId}/periods/${periodId}`
    )
    return response.data
  },

  startReview: async (
    clientId: string, 
    periodId: string, 
    request: ReviewPeriodRequest = {}
  ): Promise<ReviewPeriodResponse> => {
    const response = await api.post<ReviewPeriodResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}/review`,
      request
    )
    return response.data
  },

  finalizePeriod: async (
    clientId: string, 
    periodId: string, 
    request: FinalizePeriodRequest = {}
  ): Promise<FinalizePeriodResponse> => {
    const response = await api.post<FinalizePeriodResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}/finalize`,
      request
    )
    return response.data
  },

  lockPeriod: async (
    clientId: string, 
    periodId: string, 
    request: LockPeriodRequest
  ): Promise<LockPeriodResponse> => {
    const response = await api.post<LockPeriodResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}/lock`,
      request
    )
    return response.data
  },

  updateStatus: async (
    clientId: string,
    periodId: string,
    request: UpdatePeriodStatusRequest
  ): Promise<UpdatePeriodStatusResponse> => {
    const response = await api.patch<UpdatePeriodStatusResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}`,
      request
    )
    return response.data
  },

  getSnapshot: async (clientId: string, periodId: string): Promise<PeriodSnapshot> => {
    const response = await api.get<PeriodSnapshot>(
      `/accountant/clients/${clientId}/periods/${periodId}/snapshot`
    )
    return response.data
  },

  getAuditLogs: async (
    clientId: string, 
    periodId: string, 
    limit = 50
  ): Promise<PeriodAuditLogsResponse> => {
    const response = await api.get<PeriodAuditLogsResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}/audit-logs`,
      { params: { limit } }
    )
    return response.data
  },
}
